sync/src/main/java/tech/pegasys/artemis/sync/FetchRecentBlocksService.java

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.sync;
+
+import static tech.pegasys.artemis.util.async.FutureUtil.ignoreFuture;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.artemis.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.artemis.networking.eth2.Eth2Network;
+import tech.pegasys.artemis.service.serviceutils.Service;
+import tech.pegasys.artemis.sync.FetchBlockTask.FetchBlockResult;
+import tech.pegasys.artemis.util.async.SafeFuture;
+import tech.pegasys.artemis.util.events.Subscribers;
+
+class FetchRecentBlocksService extends Service {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private static final int MAX_CONCURRENT_REQUESTS = 3;
+  private static final Duration WAIT_FOR_PEERS_DURATION = Duration.ofSeconds(30);
+  private static final RetryDelayFunction DEFAULT_RETRY_DELAY_FUNCTION =
+      RetryDelayFunction.createExponentialRetry(2, Duration.ofSeconds(5), Duration.ofMinutes(5));
+
+  private final int maxConcurrentRequests;
+  private final Eth2Network eth2Network;
+  private final PendingPool<SignedBeaconBlock> pendingBlocksPool;
+
+  private final Map<Bytes32, FetchBlockTask> allTasks = new ConcurrentHashMap<>();
+  private final Queue<FetchBlockTask> pendingTasks = new ConcurrentLinkedQueue<>();
+  private final Collection<FetchBlockTask> activeTasks = new ConcurrentLinkedQueue<>();
+
+  private final FetchBlockTaskFactory fetchBlockTaskFactory;
+  private final RetryDelayFunction retryDelayFunction;
```

Comment:
If we're only going to use the statically declared RetryDelayFunction above, why also have a member variable instance of it?

Reply:
This pattern lets us inject mock values for testing.  But it doesn't look like I'm really doing anything special with this in the tests.  I'll look into simplifying.